### PR TITLE
Fix a variety of bugs found in dllist.c while addressing Issue #1559.

### PR DESCRIPTION
**Pull request by jmpenn**
* Added More descriptive error messages.
* Wrote google-test based unit tests for the DLL List library.
* Deleted old, incomplete test program.
* Added new function DLL_ListContainsPos which determines whether the given list contains the node at the given pos.
* DLLFind bug: Added check to determine whether a compare function has been specified, and emit an error message if it hasn’t.
* DLL_FindIndex bug: Corrected bounds checking.
* DLL_GetAt bug: Added check to ensure that the specified pos is actually in the given list.
* DLL_SetAt bug: Added check to ensure that the specified pos is actually in the given list.
* DLL_RemoveAt: corrected logic mistake in NULL ptr check. Added check to ensure that the specified pos is actually in the given list.
* DLL_InsertBefore bug: element count not correctly updated.
* DLL_InsertAfter: next and prev ptrs not correctly updated, which corrupted the list.
* DLL_GetNext bug: logic error in NULL ptr checks. Added check to ensure that the specified pos is actually in the given list.
* DLL_GetPrev bug: logic error in NULL ptr checks. Added check to ensure that the specified pos is actually in the given list.
* DLL_AddHead: Fixed NULL check logic error.
* DLL_AddTail: Fixed NULL check logic error.

### CODE REVIEW

**Comment by sharmeye** [comment]
I can think of at least one person who will be very happy with this update.  Everything looks good to me, and i like there's a unit test for this library now.

**Review by sharmeye — APPROVED** [review]
I don't see anything worrisome, looks good to me.

File: trick_source/trick_utils/trick_adt/src/dllist.c [hunk 1]
```diff
@@ -19,12 +19,13 @@ DLLIST *DLL_Create(void)
 
 /* initializes a list */
 /* return: none */
```
Comment (hchen99) [inline trick_source/trick_utils/trick_adt/src/dllist.c:21]: return: none

File: trick_source/trick_utils/trick_adt/src/dllist.c [hunk 2]
```diff
@@ -34,15 +35,34 @@ void DLL_Delete(DLLIST * list)
     }
 }
 
+/* Determine whether the given list contains the node at the given pos.
+   If it does, return 1, otherwise return false.
```
Comment (hchen99) [inline trick_source/trick_utils/trick_adt/src/dllist.c:39]: otherwise return 0 ;-)

File: trick_source/trick_utils/trick_adt/src/dllist.c [hunk 3]
```diff
@@ -34,15 +35,34 @@ void DLL_Delete(DLLIST * list)
...
+        lpos = lpos->next;
+    }
+    if (lpos == NULL) {
+        return 0;
+    }
+    return 1;
+}
 
 /* returns number of elements in list */
 /* return: count */
```
Comment (hchen99) [inline trick_source/trick_utils/trick_adt/src/dllist.c:57]: return: none

**Comment by hchen99** [comment]
Really like those descriptive error messages

File: trick_source/trick_utils/trick_adt/src/dllist.c (hunk 2 above)
Comment (jmpenn) [reply trick_source/trick_utils/trick_adt/src/dllist.c:39]: Addressed this in commit below.

**Review by hchen99 — APPROVED** [review]

**Pull request merged**